Always use estimateGas when computing forgeOrg gas

dkent600 opened the pull request:
Resolves: #342 

- always uses `estimateGas` when computing `forgeOrg` gas. No more need to manually compute the gasLimit
- caps gas limit on *any* attempt to override the default gas limit on a transaction
- rewrites the createDaoScript to use `DAO.new`
- uses AbsoluteVote for schemes (other than ContributionReward) in the createGenesisDao script

orenyodfat commented on lib/wrappers/daoCreator.ts:
```diff
@@ -99,7 +98,18 @@ export class DaoCreatorWrapper extends ContractWrapperBase {
+      controllerAddress,
+      options.tokenCap];
+
+    // always use estimateGas or will fail when using non-universal controller
```
"or will fail when using non-universal controller ? why ?

dkent600 replied:
It runs out of gas.  I don't know why the default gas limit doesn't work.

orenyodfat commented on gasLimits.js:
```diff
@@ -10,34 +10,16 @@ const arcForgeOrgGasLimit = require("./arcConstants.js").ARC_GAS_LIMIT;
   /**
-   * The gas limit used by Arc to forge a DAO with foundersInGasLimitArc founders
-   * and to `new` GenesisProtocol.
+   * The gas limit used by Arc to `new` GenesisProtocol.
```
This comment is not clear.please clarify

orenyodfat commented on gasLimits.js:
```diff
@@ -10,34 +10,16 @@ const arcForgeOrgGasLimit = require("./arcConstants.js").ARC_GAS_LIMIT;
    * Gas limit sufficient (though not necessarily optimal) for all other transactions
    * besides creating DAOs
    */
   "gasLimit_runtime": 4543760,
```
How this number is calculated ? from where it is taken ?

orenyodfat commented on gasLimits.js:
```diff
@@ -15,7 +15,8 @@ const gasLimitsConfig =
    * Gas limit sufficient (though not necessarily optimal) for all other transactions
-   * besides creating DAOs.  This is taken from the default that truffle puts in truffle.js.
+   * besides creating DAOs and instantiating a new GenesisProtocol.
+   * This is taken from the default that truffle puts in truffle.js.
```
why did you choose to use truffle default values ? why do we care about truffle default value ?
how did truffle choose this value ?

dkent600 replied:
I think the code that uses this may not be required.  It is very old code that actually predates me, and has never, that I recall, been questioned.  I will test with it removed and report back.

BTW I am also looking into removing the other setting, `gasLimit_arc`, a task which is proving to be tricky, but with both of those settings gone we could completely remove some code including the dependency on the Arc test/constants.js file.

dkent600 commented:
@orenyodfat I have removed arcConstants.js and gasLimits.js, so there is no longer a dependency on gas numbers from Arc.

The default gas limit used by truffle is now stored in defaults.json as a global configuration setting.  It is required for certain operations to succeed with enough gas.

I have updated the description in this PR to include all of these recent changes.

orenyodfat commented on config/default.json:
```diff
@@ -4,6 +4,7 @@
   "cacheContractWrappers": true,
   "logLevel": 9,
   "estimateGas": false,
+  "defaultGasLimit": 4543760,
```
why this value and not other (e.g 4543759) ?
Is there a good reason for that except that "truffle use that" ?

orenyodfat replied:
![quote-there-is-nothing-without-reason-gottfried-leibniz-126-13-19-2](https://user-images.githubusercontent.com/5994398/45596295-bd912480-b9c2-11e8-88cd-4693d3386421.jpg)

orenyodfat replied:
lets find out the reason or set it to max gas limit

orenyodfat commented on lib/contractWrapperBase.ts:
```diff
@@ -184,9 +184,7 @@ export abstract class ContractWrapperBase implements IContractWrapper {
+    const maxGasLimit = await UtilsInternal.computeMaxGasLimit();
 
     if (currentNetwork === "Ganache") {
       return maxGasLimit; // because who cares with ganache and we can't get good estimates from it
```
Gas estimation for Ganache should be the same as other networks.
In general, we should strive to execute same arc.js code regardless that networks it is connected to.(as possible as can be done). 
This line should be removed.

dkent600 replied:
Ganache returns gas estimates that are not sufficient.  At the time I wrote this code I spent a lot of time on this issue.  I just now gave it another try, but encountered the same result.  I am in the process of moving to truffle 5.0 and web3 1.0 -- maybe something will change then, I can retry when I have things working enough to test.

I think this may be a case requiring network-specific code.

orenyodfat replied:
Is this a known ganache issue ? If it is please add a link to this issue.

dkent600 replied:
https://github.com/trufflesuite/ganache-core/issues/151

orenyodfat replied:
could you please add a link to that on the code ?

orenyodfat commented on lib/contractWrapperBase.ts:
```diff
@@ -321,9 +319,9 @@ export abstract class ContractWrapperBase implements IContractWrapper {
       const gasPriceComputer = ConfigService.get("gasPriceAdjustment") as GasPriceAdjustor;
+      const web3 = await Utils.getWeb3();
 
       if (gasPriceComputer && !web3Params.gasPrice) {
```
gasPriceComputer -> gasPriceCompute ?

dkent600 replied:
fixed

orenyodfat commented on lib/migrations/2_deploy_schemes.ts:
```diff
@@ -53,7 +53,7 @@ export const arcJsDeployer = (
     const DAOToken = await Utils.requireContract("DAOToken");
-    const gasLimit = await computeMaxGasLimit(web3);
+    const gasLimit = await UtilsInternal.computeMaxGasLimit();
     const gasPrice = 10000000000; // 10 Gwei
```
const gasPrice = 10000000000; // 10 Gwei ?
from where this come from ?

dkent600 replied:
The scripts doubles that in a few specific cases where it doesn't suffice.

orenyodfat replied:
fix 10 Gwei is a workaround.
please set the appropriate gas price needed.
You can read that from eth gas station ...(alchemy done that)

dkent600 replied:
Yeah, I was thinking of that too.  Would you mind waiting on this until I have Truffle 5 working?  As you know, they have lots of migration changes and I'd like to code this in the context of those changes.  I'm doing the work right now of upgrading to Truffle 5.

orenyodfat replied:
please open an issue for that.

dkent600 commented:
New tickets:

https://github.com/daostack/arc.js/issues/358
https://github.com/daostack/arc.js/issues/357
https://github.com/daostack/arc.js/issues/359

@orenyodfat
